### PR TITLE
scheduler(ticdc): inline secondary to captures map 

### PR DESCRIPTION
**Pull request by overvenus**
### What problem does this PR solve?

Issue Number: ref #4757 

### What is changed and how it works?

Inline `Secondary` to `ReplicationSet.Captures` and remove `Secondary` in `ReplicationSet`.

#### Tests 

 - Unit test
 - Integration test

##### Will it cause performance regression or break compatibility?

No.

##### Do you need to update user documentation, design documentation or monitoring documentation?

No.

### Release note 

```release-note
None
```

### CODE REVIEW

File: cdc/scheduler/internal/v3/replication/replication_set.go [hunk 1]
```diff
@@ -90,16 +90,49 @@ func (r ReplicationSetState) MarshalJSON() ([]byte, error) {
...
-	// Primary set indicate that the table is replicating on this capture
+	// Primary is the capture ID that is currently replicating the table.
 	Primary model.CaptureID
-	// Secondary set indicate that the table is preparing on this capture
-	Secondary  model.CaptureID
-	Captures   map[model.CaptureID]struct{}
+	// Captures is a map of captures that has the table replica.
+	// NB: Invariant, 1) at most one primary, 2) primary capture must be in
+	//     CaptureRolePrimary.
+	Captures   map[model.CaptureID]CaptureRole
```
Comment (3AceShowHand) [inline cdc/scheduler/internal/v3/replication/replication_set.go:135]: I'm quite curious, Why we have `Captures` field, which only have 2 items? 

Why not just directly use `Primary` and `Secondary` fields ?

File: cdc/scheduler/internal/v3/replication/replication_set.go [hunk 2]
```diff
@@ -195,6 +241,75 @@ func NewReplicationSet(
 	return r, nil
 }
 
+func (r *ReplicationSet) hasRole(role CaptureRole) bool {
+	_, has := r.getRole(role)
+	return has
+}
+
+func (r *ReplicationSet) isInRole(captureID model.CaptureID, role CaptureRole) bool {
```
Comment (3AceShowHand) [inline cdc/scheduler/internal/v3/replication/replication_set.go:249]: just name it `inRole` should be ok.

File: cdc/scheduler/internal/v3/replication/replication_set.go [hunk 3]
```diff
@@ -374,16 +487,17 @@ func (r *ReplicationSet) pollOnPrepare(
 				zap.Stringer("tableState", input),
 				zap.String("captureID", captureID),
 				zap.Any("replicationSet", r))
-			r.Primary = ""
-			delete(r.Captures, captureID)
+			r.clearPrimary()
 			return nil, false, nil
-		} else if r.Secondary == captureID {
+		} else if r.isInRole(captureID, CaptureRoleSecondary) {
```
Comment (3AceShowHand) [inline cdc/scheduler/internal/v3/replication/replication_set.go:492]: `else if` can be simplified to `if` directly.

File: cdc/scheduler/internal/v3/replication/replication_set.go (hunk 1 above)
Comment (overvenus) [reply cdc/scheduler/internal/v3/replication/replication_set.go:135]: I'm planning to support adding more replicas to tables, so TiCDC can handle unplanned shutdown.

File: cdc/scheduler/internal/v3/replication/replication_set.go (hunk 2 above)
Comment (overvenus) [reply cdc/scheduler/internal/v3/replication/replication_set.go:249]: The usual convention to name methods that return boolean is to prefix verbs such as 'is' or 'has' to the predicate as a question, or use the predicate as an assertion.

File: cdc/scheduler/internal/v3/replication/replication_set.go (hunk 3 above)
Comment (overvenus) [reply cdc/scheduler/internal/v3/replication/replication_set.go:492]: I am planning to add a new role in later PR, called `Standby` for handling unplanned shutdown.

**Review by asddongmen — APPROVED** [review]

**Review by 3AceShowHand — APPROVED** [review]

**Pull request merged**